clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -1689,6 +1692,25 @@ mlir::Location CIRGenModule::getLocForFunction(const clang::FunctionDecl *FD) {
   return theModule->getLoc();
 }
 
+void CIRGenModule::setCodegenAttributesForFunc(FuncOp f,
+                                         const clang::FunctionDecl *FD) {
+  // Track whether we need to add the optnone attribute,
+  // starting with the default for this optimization level.
+  bool ShouldAddOptNone =
+      !codeGenOpts.DisableO0ImplyOptNone && codeGenOpts.OptimizationLevel == 0;
+  ShouldAddOptNone &= !FD->hasAttr<MinSizeAttr>();
+  ShouldAddOptNone &= !FD->hasAttr<AlwaysInlineAttr>();
+
+  if ((ShouldAddOptNone || FD->hasAttr<OptimizeNoneAttr>())) {
+    f->setAttr(builder.getStringAttr("cir.optnone"),
```

Comment:
I think we should go for a more principled approach when introducing "LLVM forward-only attributes". Feels "loose" to introduce those in the wild, without a central place to track them. We might want to look at those on verifiers and other things in the future.

As I mentioned this before in our chats, I have two suggestion on how to proceed:
1. Make the attribute first-class citizen (e.g. like we do for linkage) - I can see CIR analysis looking at both noinline and optnone, but I'd be fine if they are LLVM forwardable-only for now. And we can latter move to first-class later when we need.
2. Have a wrap for "LLVM forward-only attributes" attributes: create a `LLVMAttrFwdAttr` (perhaps with a better name), add enums for each one we introduce, and set them here.

Thoughts? Suggestions?

Reply:
Test!

Reply:
> Anyways, I thought maybe it is better to make a standalone CIR attribute defined in CIRAttrs.td, instead of using hard-coded names here and later on in LLVM translation.

Yes, this is my main point from previous comment.

> However my concern is there are way more attributes to be added and each of will need a specific attribute which may flood CIRAttrs.td. But looking at how Clang and LLVM attributes are handled, they do have their own standalone definition, so maybe that's the way to go.

Better than defined at insert time :) 

> `LLVMAttrFwdAttr` sounds a bit tied to LLVM.

It was an example, thanks for expanding on the idea. What I'm trying to convey is that given the number of different attributes (as you noted in your comment) we might wanna group things CIR doesn't "understand" and effectively just a passthrough.

> Some attributes are quite general, such as optnone and noinline. How about about naming them `CodegenAttr` or `BackendAttr`? 

I guess we could start simple with a CIRFuncAttr that just wraps a ArrayAttr list, which in code could look something like `#cir.fnattr<#cir.noinline, #cir.optnone>`, we could then later use aliases per func so that the annotations are terse and easier to read. In the future we can then refine those if it makes sense? (e.g. by making then relative to backend, optimization, etc). This can probably reach out current goal of unifying those somewhere and get you unblocked. My feeling is that as we add more it will make more sense how we want to group them, perhaps it's too early and we might not wanna overthink this right now.

> BTW, what does a wrapper attribute look like? Do we an an example in CIR code base?

It was just an idea for encapsulation. Take `CastKind` for instance, it's composed of individual `CK_.*` enums.

Reply:
> I guess we could start simple with a CIRFuncAttr that just wraps a ArrayAttr list, which in code could look something like #cir.fnattr<#cir.noinline, #cir.optnone>

Sounds good. I was experimenting with a DictionaryAttr. So to conclude, we should probably go with a ArrayAttr or DictionaryAttr of a set of custom-defined CIR_Attr attributes instead of a raw named attribute, each of which is for optnone, noinline ... Does that sound good to you?

Reply:
Yep!

Reply:
Please see if the second version is expected. There are other test failures due to new funcAttr I will have to clean up.